core/trino-main/src/main/java/io/trino/server/security/oauth2/ScribeHttpClient.java

```diff
@@ -0,0 +1,211 @@
+    {
+        return execute(userAgent, headers, httpVerb, completeUrl, createStaticBodyGenerator(bodyContents, UTF_8));
+    }
+
+    @Override
+    public Response execute(String userAgent, Map<String, String> headers, Verb httpVerb, String completeUrl, File bodyContents)
+            throws InterruptedException, ExecutionException, IOException
+    {
+        throw new UnsupportedOperationException("Not implemented");
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        httpClient.close();
+    }
+
+    private Response execute(
+            String userAgent,
+            Map<String, String> headers,
+            Verb httpVerb,
+            String completeUrl,
+            BodyGenerator bodyGenerator)
+    {
+        ImmutableMultimap.Builder<String, String> requestHeaders = ImmutableMultimap.<String, String>builder()
+                .putAll(DEFAULT_HEADERS)
+                .putAll(headers.entrySet());
+        if (userAgent != null) {
+            requestHeaders.put(USER_AGENT, userAgent);
+        }
+        Request request = Request.builder()
+                .setMethod(httpVerb.name())
+                .setUri(URI.create(completeUrl))
+                .setBodyGenerator(bodyGenerator)
+                .addHeaders(requestHeaders.build())
+                .build();
+        try {
+            return httpClient.execute(request, new ConvertingResponseHandler());
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static class ConvertingResponseHandler
+            implements ResponseHandler<Response, Exception>
+    {
+        private static final Joiner HEADER_JOINER = Joiner.on(", ");
+
+        @Override
+        public Response handleException(Request request, Exception exception)
+                throws Exception
+        {
+            throw exception;
+        }
+
+        @Override
+        public Response handle(Request request, io.airlift.http.client.Response response)
+                throws Exception
+        {
+            return convertResponse(response);
+        }
+
+        private static Response convertResponse(io.airlift.http.client.Response response)
+                throws IOException
+        {
+            int code = response.getStatusCode();
+            Map<String, String> headers = response.getHeaders().asMap()
+                    .entrySet()
+                    .stream()
+                    .collect(toImmutableMap(
+                            e -> e.getKey().toString(),
+                            e -> HEADER_JOINER.join(e.getValue())));
+            return new Response(
+                    code,
+                    HttpStatus.fromStatusCode(code).toString(),
+                    headers,
+                    // airlift closes the stream before Scribe is able to even use it
+                    CharStreams.toString(new InputStreamReader(response.getInputStream())));
```

Comment:
This creates a `InputStreamReader` using the JVM default charset. There are two reasonable choices

1. Use the character set declared in the response. This is the correct way.
2. Assume the response is UTF-8 and use that. This is what their `Response` class does anyway if you pass it an `InputStream`, which is not great, but it means we aren't any worse.

I looked through their library to find where the response body is used and code seems questionable. In the default `OAuth2AccessTokenJsonExtractor`, it doesn't check that the `Content-Type` was JSON and just assumes that it's right if it got a 200 response code.

Reply:
I've used the implementation from: `com.github.scribejava.core.model.Response#parseBodyContents` which also assumes that the input stream is UTF-8.